src/io_uring.c

```diff
@@ -0,0 +1,91 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+#include <string.h>
+
+#include "zmalloc.h"
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+struct io_uring_context {
+    struct io_uring ring;
+    size_t queue_len;
+};
+
+io_uring_context *createIOUring(void) {
+    io_uring_context *uring_context = zmalloc(sizeof(struct io_uring_context));
+    uring_context->queue_len = 0;
+    struct io_uring_params params;
+    memset(&params, 0, sizeof(params));
+    /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+     * point to the shared memory containing the io_uring queues.
+     * On failure -errno is returned. */
+    if (io_uring_queue_init_params(IO_URING_DEPTH, &uring_context->ring, &params) < 0) return NULL;
+    return uring_context;
+}
+
+/* Submit fdatasync request to io_uring. */
+int ioUringPrepFsyncAndSubmit(io_uring_context *ring_context, int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring_context->ring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
+    ring_context->queue_len++;
+    int ret = io_uring_submit(&ring_context->ring);
+    if (ret < 0) {
+        errno = -ret;
+        return IO_URING_ERR;
+    }
+    return IO_URING_OK;
+}
+
+int ioUringWaitFsyncBarrier(io_uring_context *ring_context) {
+    struct io_uring_cqe *cqe;
+    while (ring_context->queue_len) {
+        int ret = io_uring_wait_cqe(&ring_context->ring, &cqe);
+        if (ret == 0) {
+            if (cqe->res < 0) {
+                errno = -cqe->res;
+                return IO_URING_ERR;
```

Comment:
> Will aborting in the middle leak the CQEs and SQEs?

Ditto, based on the scenario and existing implementation now, it only have one fsync job, SQEs will be empty once io_uring_submit finished and when we enter this barrier function, the CQEs should have only one entry. 

> I think we need to process all pending CQEs but then it becomes tricky to handle errors, when there are multiple failed CQEs. I wonder if we should just return the number of successfully completed requests. It would be challenging however to bubble up the errnos associated with the failed requests, with the abstraction model in this PR. So an alternative would be to serverLog the errors in this function only and not propagate them back to the caller.

Maybe we can implement this in https://github.com/valkey-io/valkey/pull/112? This scenario will have multiple write jobs.

Reply:
>based on the scenario and existing implementation now, it only have one fsync job, SQEs will be empty once io_uring_submit finished and when we enter this barrier function, the CQEs should have only one entry.

I think this assumes the caller will alternate between submitting fsync jobs and waiting on the barrier. But I don't see this assumption getting explicitly enforced in io_uring.c. Implicitly relying on caller behavior can lead to unnecessary coupling and potential errors if the caller deviates from this pattern.